ICS20-2: Add Token Forwarding ability to ICS20-2 (#1090)
cosmos/ibc: spec/app/ics-020-fungible-token-transfer/README.md

  sender: string
  receiver: string
  memo: string
  forwardingPath: []string // a list of portID/channelID pairs determining where the tokens must be forwarded next
}

interface Token {
  denom: string // base denomination

>   trace: []string

Reviewer: it would be helpful if maybe we clarified the naming here, between `trace` and `path` -- below we understand that traces are only the port and channel pairs, but in ie: the `DenomTrace` struct we understand that concept of trace as containing the path AND base denom.

Reviewer: there's also quite a lot of overlap between DenomTrace and Token, the only difference seems to be that the Token trace is structured a bit differently than the trace on DenomTrace (port/channel pairs v one long string), and that there's an amount field. would we maybe want to embed the DenomTrace struct into Token and then have those methods be available there?

Reviewer 2: Thanks @charleenfei for raising this concern. The approach you propose is pretty similar to the one that our colleagues at informal systems have used to model the ics20v1 in quint. I personally like it.

  sender: string
  receiver: string
  memo: string
  forwardingPath: ForwardingInfo // a struct containing the list of next hops, determining where the tokens must be forwarded next, and the memo for the final hop
}

interface ForwardingInfo {
  hops: []Hop
  memo: string,

> }

Reviewer 3: Personally, I feel like the memo in `FungibleTokenPacketData` should cover this, and that we should not include a memo field in this data structure.

Reviewer 3: There would also be overlap between `ForwardingInfo` / `Hop` and the `Trace` type which includes the port id and channel id tuple.

Reviewer 3: > Personally, I feel like the memo in FungibleTokenPacketData should cover this, and that we should not include a memo field in this data structure.

I understand the concern here is preventing unintentional middleware invocation from source chain or intermediary recving chain, but also thought about what if you can rather have a memo attached to each individual hop. 

Spoke about this with @colin-axner and fine to keep this structure as introducing this on a per-hop basis is opening up a big rabbit hole of potential issues.

```typescript
interface ModuleState {
  channelEscrowAddresses: Map<Identifier, string>

>   channelForwardingAddresses: Map<Identifier, string>

Reviewer 2: I agree, we could use directly channelEscrowAddresses.  The channelEscrowAddresses mapping is populated during `onChainOpenTry` at line 181. Instead the channelForwardingAddresses is never set. Thus the read we do at line 331 may be illegal.

>   channelForwardingAddresses: Map<Identifier, string>

Reviewer 3: note https://github.com/cosmos/cosmos-sdk/issues/7737#issuecomment-735671951

>   channelForwardingAddresses: Map<Identifier, string>

Reviewer 4: The forwarding address is mainly just used as pass through storage based on how ics20 is constructed in the non-forwarding case. I think the spec just needs to specify that this address should be an address the transfer module controls. I don't think it should reuse the escrow addresses

>   channelForwardingAddresses: Map<Identifier, string>

Reviewer 4: Unless I am missing something, I can't think of a reason why you need channel isolated forwarding addresses https://github.com/cosmos/ibc-go/issues/6561#issuecomment-2176403684

>   channelForwardingAddresses: Map<Identifier, string>

Author: The reason I decided to split it off to a separate address is just an additional security blanket. The balance of the escrow addresses is security critical. If the balance ends up in an invalid state, that can be a very painful issue to unwind.

It should be completely fine if there's no bugs. But if there's a bug in forwarding, i figure having a separate account might make the impact a bit more isolated.

Of course it is receiving from and sending back to the escrow addresses, so if there's issues there it could still be a problem for escrow account balances.

I'm ok either way. I just figured to start out with a separate account for clarity and avoiding having the escrow account representing more than one thing semantically.

We can use the same account as well as @colin-axner suggests

  ],
  sender: cosmosexampleaddr1,
  receiver: cosmosexampleaddr2,
  memo: "",

>   forwardingPath: {[{"transfer", "channel-7"}, {"transfer", "channel-13"}], , "swap: {...}"}, // provide hops in order and the memo intended for final hop

Reviewer 2: Is the double `,` close to swap in line 114 intentional?  Should it be `forwardingPath: {[{"transfer", "channel-7"}, {"transfer", "channel-13"}], "swap: {...}"}, // provide hops in order and the memo intended for final hop` ?

  sender: string,
  receiver: string,
  memo: string,
  forwardingPath: ForwardingInfo,
  sourcePort: string,
  sourceChannel: string,
  timeoutHeight: Height,
  timeoutTimestamp: uint64, // in unix nanoseconds
): uint64 {
  // memo and forwardingPath cannot both be non-empty

>   abortTransactionUnless(memo != "" && forwardingPath != nil)

Reviewer 5: Actually, couldn't we get rid of one of of the two memos? If the memo supposed to trigger an action on the final destination of the tokens, why cannot we have something like this?

```
interface FungibleTokenPacketDataV2 {
  tokens: []Token
   sender: string
   receiver: string
   memo: string
   forwardingPath: ForwardingInfo
}

interface ForwardingInfo {
   hops: []Hop
}
```

But then it would not be possible to trigger any action on intermediary hops. Is that desired behaviour?

       amount: packet.amount
     }
     tokens = []Token{token}
     receiver = packet.receiver
  } else if transferVersion == "ics20-2" {
    FungibleTokenPacketDataV2 data = UnmarshalJSON(packet.data)
    tokens = data.tokens
    // if we need to forward the tokens onward
    // temporarily send to the channel escrow address of the intended receiver

>     if len(forwardingPath) > 0 {

Reviewer 2: For consistency we should use `len(forwardingPath)>0` or `forwardingPath != nil` in both lines 330 and 398.

>       receiver = channelForwardingAddresses[packet.destinationChannel]

Reviewer 2: Could be ok. If we remove the `channelForwardingAddresses` and use `channelEscrowAddresses` we could do directly: 

`receiver = channelEscrowAddresses[packet.destinationChannel]`

>       receiver = channelForwardingAddresses[packet.destinationChannel]

Reviewer 5: I still think that this should be:

```
nextPort, nextChannel = split(forwardingPath[0], "/")
channelEscrowAddresses[nextChannel]
```

because `packet.destinationChannel` is the same as the source channel where tokens where sent from, and therefore `channelEscrowAddresses[packet.destinationChannel]` would be equal to the escrow address where the tokens were locked when they left the chain. 

I hope this makes sense!

>       receiver = channelForwardingAddresses[packet.destinationChannel]

Author: Since we are using different addresses from the escrow this should be fine so long as we stay consistent. Do you see a bug in the logic, or just a preference?

>     }

Reviewer 2: ```suggestion
    } else{
              receiver=data.receiver
              }
```

  } else {
    // should never be reached as transfer version must be negotiated to be either
    // ics20-1 or ics20-2 during channel handshake
    abortTransactionUnless(false)
  }

  assert(packet.sender !== "")

>   assert(receiver !== "")

Reviewer 2: Note that this assertion may fail if we are in the case where `transferVersion == ics-20-2 && forwardingPath is nil` because we are not setting the `receiver` variable. If I got it correctly this would be the case where we are the destChain and we don't need to keep forwarding. Thus we should handle this scenario setting a proper receiver to allow lines 362 or 374 to execute properly.

      onChainTrace = append([]string{prefix}, token.trace...)
      onChainDenom = constructOnChainDenom(onChainTrace, token.denom)
      // sender was source, mint vouchers to receiver (assumed to fail if balance insufficient)
      err = bank.MintCoins(receiver, onChainDenom, token.amount)
      if (err !== nil) {
        ack = FungibleTokenPacketAcknowledgement{false, "mint coins failed"}
        // break out of for loop on first error
        break
      }
    }

    // add the received token to the received tokens list
    recvToken = Token{
      denom: token.denom,
      trace: onChainTrace,
      amount: token.amount,
    }
    receivedTokens = append(receivedTokens, recvToken)
  }

  // if there is an error ack return immediately and do not forward further
  if !ack.Success() {
    return ack
  }

  // if acknowledgement is successful and forwarding path set
  // then start forwarding
  if forwardingPath != nil {
    nextPort, nextChannel = split(forwardingPath[0], "/")
    // send the tokens we received above to the next port and channel
    // on the forwarding path
    // and reduce the forwardingPath by the first element

>     nextSequence = sendFungibleTokens(

Reviewer 2: ```suggestion
    nextPacketSequence = sendFungibleTokens(
```

>     nextSequence = sendFungibleTokens(

Reviewer 5: in this call to `sendFungibleTokens`, when the chain is source of the tokens, we do a bank transfer where sender and receiver are the same address (i.e. the escrow account on the source channel). Just mentioning, in case there's a way to avoid the extra bank transfer.

      tokens: receivedTokens,
      sender: receiver,

>       receiver: data.receiver,

Reviewer 2: Here receiver is a global variable which is either set at line 324 or 331. Additionally we verify receiver is not empty at line 340. I agree we should use directly receiver.

      memo,
      forwardingPath[1:],
      nextPort,
      nextChannel,
      Height{},
      currentTime() + DefaultHopTimeoutPeriod,
    )
    // store packet for future sending ack

>     store.set(packetForwardPath(nextPort, nextChannel, nextPacketSequence), packet)

Reviewer 2: should we rename the store as a ProvableStore?

>     store.set(packetForwardPath(nextPort, nextChannel, nextPacketSequence), packet)

Reviewer 5: I think it should be `privateStore` instead, since we don't need to prove anything against this path, right? The packet that is stored in this key is just auxiliary data to correctly execute refund logic, right?

>     store.set(packetForwardPath(nextPort, nextChannel, nextPacketSequence), packet)

Reviewer 2: Yes, you're right! The prevPacket is used to call the revertInFlightChanges which works locally in the calling chain.

>        amount: packet.amount

Reviewer 2: should this be data.amount instead?

     }
     tokens = []Token{token}
     receiver = data.receiver
  } else if transferVersion == "ics20-2" {
    FungibleTokenPacketDataV2 data = UnmarshalJSON(packet.data)
    tokens = data.tokens
    // if we need to forward the tokens onward
    // overwrite the receiver to temporarily send to the channel escrow address of the intended receiver
    if len(forwardingPath.hops) > 0 {
      // memo must be empty
      abortTransactionUnless(memo == "")
      if channelForwardingAddress[packet.destinationChannel] == "" {
        channelForwardingAddress[packet.destinationChannel] = newAddress()
      }
      receiver = channelForwardingAddresses[packet.destinationChannel]
      finalReceiver = data.receiver
    } else {
      receiver = data.receiver
    }
  } else {
    // should never be reached as transfer version must be negotiated to be either
    // ics20-1 or ics20-2 during channel handshake
    abortTransactionUnless(false)
  }

>   assert(packet.sender !== "")

Reviewer 2: Should this be data.sender?

    tokens = data.tokens
    // if we need to forward the tokens onward
    // overwrite the receiver to temporarily send to the channel escrow address of the intended receiver
    if len(forwardingPath.hops) > 0 {
      // memo must be empty

>       abortTransactionUnless(memo == "")

Reviewer 5: Yes, also `forwardingPath`.

    channel = provableStore.get(channelPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID))
    if channel.version != "ics20-2" && len(forwardingPath.hops) > 1 {
      ack = FungibleTokenPacketAcknowledgement(false, "next hop in path cannot support forwarding onward")
      return ack
    }
    memo = ""
    nextForwardingPath = ForwardingPath{
      hops: forwardingPath.hops[1:]
      memo: forwardingPath.memo
    }
    if forwardingPath.hops == 1 {
      // we're on the last hop, we can set memo and clear
      // the next forwardingPath
      memo = forwardingPath.memo
      nextForwardingPath = nil
    }
    // send the tokens we received above to the next port and channel
    // on the forwarding path
    // and reduce the forwardingPath by the first element
    nextPacketSequence = sendFungibleTokens(
      receivedTokens,
      receiver, // sender of next packet
      finalReceiver, // receiver of next packet
      memo,
      nextForwardingPath,
      forwardingPath.hops[0].portID,
      forwardingPath.hops[0].channelID,
      Height{},
      currentTime() + DefaultHopTimeoutPeriod,
    )
    // store packet for future sending ack

>     privateStore.set(packetForwardPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID, nextPacketSequence), packet)

Reviewer 2: Note that we have an incompatibility in the number of parameters used with the `packetForwardPath` in the : 
- `OnRecvPacket`  :  we use `packetForwardPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID, nextPacketSequence)`. Here we include nextPacketSequence as a key for the path. 

- `onAcknowledgePacket`&& `onTimeoutPacket` :  we use `packetForwardPath(packet.sourcePort, packet.sourceChannel)` without using the nextPacketSequence as a discriminator for the key. 

The result is that the previous packet in `onAcknowledgePacket`&& `onTimeoutPacket` is always nil. 

Since, unless we store it somewhere, should be not possible to retrieve the nextPacketSequence in `onAcknowledgePacket` && `onTimeoutPacket`, a couple of possible solutions could be: 
1. remove nextPacketSequence in the key for packetForwardPath, 
2. remove nextPacketSequence in the key for packetForwardPath, but save it together with packet in store. 

Do you see a possible better approach?

  if !(acknowledgement.success) {
    refundTokens(packet)
  }

  // check if the packet was sent is from a previously forwarded packet

>   prevPacket = store.get(packetForwardPath(packet.sourcePort, packet.sourceChannel))

Reviewer 2: Same here, store should be provableStore?

      prefix = "{packet.destPort}/{packet.destChannel}/"
      onChainTrace = append([]string{prefix}, token.trace...)
      onChainDenom = constructOnChainDenom(onChainTrace, token.denom)
      // sender was source, mint vouchers to receiver (assumed to fail if balance insufficient)
      err = bank.MintCoins(receiver, onChainDenom, token.amount)
      if (err !== nil) {
        ack = FungibleTokenPacketAcknowledgement{false, "mint coins failed"}
        // break out of for loop on first error
        break
      }
    }

    // add the received token to the received tokens list
    recvToken = Token{
      denom: token.denom,
      trace: onChainTrace,
      amount: token.amount,
    }
    receivedTokens = append(receivedTokens, recvToken)
  }

  // if there is an error ack return immediately and do not forward further
  if !ack.Success() {
    return ack
  }

  // if acknowledgement is successful and forwarding path set
  // then start forwarding
  if len(forwardingPath.hops) > 0 {
    //check that next channel supports token forwarding
    channel = provableStore.get(channelPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID))

>     if channel.version != "ics20-2" && len(forwardingPath.hops) > 1 {

Reviewer 6: Does all the previous minting / transfer need to be reverted here, or is it implicit?

Reviewer 5: Yes, it's kinda implicit: if an error ack is returned then the datagram handler should not commit any state changes. But I agree, that maybe it would be good to mention this explicitly in the specs (maybe in ICS26? I cannot remember from the top of my head if we already do that...

      ack = FungibleTokenPacketAcknowledgement(false, "next hop in path cannot support forwarding onward")
      return ack
    }
    memo = ""
    nextForwardingPath = ForwardingPath{
      hops: forwardingPath.hops[1:]
      memo: forwardingPath.memo
    }
    if forwardingPath.hops == 1 {
      // we're on the last hop, we can set memo and clear
      // the next forwardingPath
      memo = forwardingPath.memo
      nextForwardingPath = nil
    }
    // send the tokens we received above to the next port and channel
    // on the forwarding path
    // and reduce the forwardingPath by the first element
    packetSequence = sendFungibleTokens(
      receivedTokens,
      receiver, // sender of next packet
      finalReceiver, // receiver of next packet
      memo,
      nextForwardingPath,
      forwardingPath.hops[0].portID,
      forwardingPath.hops[0].channelID,
      Height{},
      currentTime() + DefaultHopTimeoutPeriod,
    )
    // store packet for future sending ack

>     privateStore.set(packetForwardPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID, packetSequence), packet)

Reviewer 4: only store packetID, storing the other packet fields in unnecessary

Author: Need to change the function signature of writeAcknowledgement which I didn't want to do in this PR. But agree we should probs do for release

Author: But the acknowledgement handler needs the full packet. The relayers currently grab the packet from the writeAcknowledgement event emissions. So we would need to think a bit more about how to handle relaying in the async ack case.

function onTimeoutPacket(packet: Packet) {
  // the packet timed-out, so refund the tokens

>   refundTokens(packet)

Reviewer 5: I think we need some `if` statement is around `refundTokens`, otherwise if `prevPacket` is not `nil`, aren't we executing refund logic twice?

Alternatively, maybe there is a way in which we can execute `refundTokens` always and if `prevPacket` is not nil, then we execute additional logic in `revertInFlightChanges` (but this logic could be simplified from its current form).

Author: i don't believe so. 

So as an example, during a receive in an intermediate chain:

We might mint tokens to `receiverAddrA` and then sending from `receiverAddrA` to the forward escrow account.

The refundTokens will simply send from the forwardEscrowAccount to `receiverAddrA` whereas revertInflightChanges will burn the tokens in `receiverAddrA`'s account.

So the intermediate receive involves a **receive** and then a **send**.

refundTokens reverts the `send`

and `revertInflightPacket` reverts the receive.

Hopefully that's useful context to verify the logic of `revertInflightChanges`

Reviewer 5: I think there are two cases on which executing logic on both `refundTokens` and `revertInflightPacket` will leave the middle chain in an undesirable state in case of an error acknowledgement.

### When middle hop is not source when sending and source when receiving

In `onRecvPacket` vouchers are transferred from escrow account to forwarding account, then in `sendFungibleTokens` vouchers in forwarding account are burnt. In `onAcknowledgePacket` in `refundTokens` vouchers are minted to forwarding account and in `revertInflightPacket` vouchers are minted on escrow account. End state: vouchers in forwarding account.

## When middle hop is not source when sending and not source when receiving

In `onRecvPacket` vouchers are minted to forwarding account, then in `sendFungibleTokens` vouchers in forwarding account are burnt. In `onAcknowledgePacket` in `refundTokens` vouchers are minted to forwarding account and in `revertInflightPacket` nothing is done. End state: vouchers in forwarding account.

I would appreciate if you can take a look at [the diagrams above](https://github.com/cosmos/ibc/pull/1090#pullrequestreview-1981084855), the bottom two diagrams show these two situations, in case I am missing something...

Reviewer 2: Hey guys, in this [PR](https://github.com/cosmos/quint/pull/3) we have the test of all the 4 scenarios depicted above regarding error acknowledgements. Assuming the translation from 'spec' to 'quint' is correct, the specifications have been proven accurate during the execution of error acknowledgments with no undesirable ending state detected. Reviews would be much appreciated. 

For reference I post here  
- [new testing scenario diagrams](https://excalidraw.com/#json=imWPrnjVUYGRX0ruIxjii,bWrrA-vuQx43Sd7Z-pSB2w), adapted to existing nomenclature present in the `ics20.qnt` specification.   
- [the state transition tables for every scenario](https://docs.google.com/spreadsheets/d/1XCI6aNBfqDYhdOpzg0tFxaXwWCPRLaRXgiuPzvePRjk/edit#gid=2106405052)

Potential next steps to ensure the correctness of this spec would be: 
- Adding timeout testing scenarios (even if, since reversion logic is correct, everything should be ok) 
- Add multidenom logic and tests.   

Again thanks @crodriguezvega and @AdityaSripal for the great work!

  if !(acknowledgement.success) {
    refundTokens(packet)
  }

  // check if the packet that was sent is from a previously forwarded packet
  prevPacket = privateStore.get(packetForwardPath(packet.sourcePort, packet.sourceChannel))

  if prevPacket != nil {
    if acknowledgement.success {
      FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{true, "forwarded packet succeeded"}
      handler.writeAcknowledgement(
        prevPacket,
        ack,
      )
    } else {
      // the forwarded packet has failed, thus the funds have been refunded to the forwarding address.
      // we must revert the changes that came from successfully receiving the tokens on our chain
      // before propogating the error acknowledgement back to original sender chain
      revertInFlightChanges(packet, prevPacket)
      // write error acknowledgement
      FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{false, "forwarded packet failed"}
      handler.writeAcknowledgement(
        prevPacket,
        ack,

>       )

Reviewer 2: ```suggestion
      )}
```
We are missing a closing bracket.

function onAcknowledgePacket(
  packet: Packet,
  acknowledgement: bytes) {
  // check if the packet that was sent is from a previously forwarded packet
  prevPacket = privateStore.get(packetForwardPath(packet.sourcePort, packet.sourceChannel, packet.sequence))

  if prevPacket != nil {

>     if acknowledgement.success {

Reviewer 6: It' not really specified here or else where, as far as I can see, but is there any point in time where it is safe for the privateStore to remove these packets?

Reviewer 5: Good point! Maybe it's possible to delete the `prevPacket` after `OnAcknowledgePacket` or `OnTimeoutPacket` have executed. What do you think, @AdityaSripal?

Author: Yes it is, added it in now

```typescript
function onTimeoutPacket(packet: Packet) {
  // check if the packet was sent is from a previously forwarded packet
  prevPacket = privateStore.get(packetForwardPath(packet.sourcePort, packet.sourceChannel))

  if prevPacket != nil {
    // the forwarded packet has failed, thus the funds have been refunded to the forwarding address.
    // we must revert the changes that came from successfully receiving the tokens on our chain
    // before propogating the error acknowledgement back to original sender chain
    revertInFlightChanges(packet, prevPacket)
    // write error acknowledgement

>     FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{false, "forwarded packet failed"}

Reviewer 5: Probably this is fine, but just wanted to comment that in case of time out, the middle chains will propagate an error ack back to the sending chain. Would there be any implications for applications or middleware on the sending chain if the packet is error acknowledged instead of timed out?

Reviewer 5: Maybe change the message to make it more specific?

```suggestion
    FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{false, "forwarded packet timed out"}
```

}
```

```typescript
// revertInFlightChanges reverts the receive packet and send packet
// that occurs in the middle chains during a packet forwarding
// If an error occurs further down the line, the state changes
// on this chain must be reverted before sending back the error acknowledgement
// to ensure atomic packet forwarding
function revertInFlightChanges(sentPacket: Packet, receivedPacket: Packet) {
  // the token on our chain is the token in the sentPacket
  for token in sentPacket.tokens {
    // check if the packet we sent out was sending as source or not
    // in this case we escrowed the outgoing tokens
    if isSource(sentPacket.sourcePort, sentPacket.sourceChannel, token) {
      // check if the packet we received was a source token for our chain
      if isSource(receivedPacket.destinationPort, receivedPacket.desinationChannel, token) {

>         // receive sent tokens from the received escrow to the forward escrow account

Reviewer 2: ```suggestion
        // Previous Actions: 
                //1. Receive tokens from the receivedEscrowAccount 
                //2. Send token to forwardEscrowAccount
```

>         // so we must send the tokens back from the forward escrow to the original received escrow account

Reviewer 2: ```suggestion
        // Reverse Actions: 
        //1. Send tokens back from forwardEscrowAccount to the original receivedEscrowAccount
```

        forwardEscrow = channelEscrowAddresses[sentPacket.sourceChannel]
        reverseEscrow = channelEscrowAddresses[packet.destChannel]
        bank.TransferCoins(forwardEscrow, reverseEscrow, token.denom, token.amount)
      } else {

>         // receive minted vouchers and sent to the forward escrow account

Reviewer 2: ```suggestion
        // Previous Actions: 
                //1. Received minted vouchers 
                //2. Send Vouchers to forwardEscrow Account. 
```

>         // so we must remove the vouchers from the forward escrow account and burn them

Reviewer 2: ```suggestion
        // Reverse Actions: 
        //1. Burn Vouchers from the forwardEscrowAccount
```

        bank.BurnCoins(forwardEscrow, token.denom, token.amount)
      }
    } else {

>       // in this case we burned the vouchers of the outgoing packets

Reviewer 2: ```suggestion
      // Previous Actions: 
```

      // check if the packet we received was a source token for our chain
      // in this case, the tokens were unescrowed from the reverse escrow account
      if isSource(receivedPacket.destinationPort, receivedPacket.desinationChannel, token) {

>         // in this case we must mint the burned vouchers and send them back to the escrow account

Reviewer 2: ```suggestion
        // // Previous Actions: 
                //1. Burn outgoing packet Vouchers
                //2. Unescrow from reverseEscrowAccount
                //Reverse Actions: 
                //1. Mint Vouchers to reverseEscrowAccount
```

>         bank.MintCoins(reverseEscrow, token.denom, token.amount)

Reviewer 5: `reserveEscrow` is not defined in this scope.

>         bank.MintCoins(reverseEscrow, token.denom, token.amount)

Reviewer 2: ```suggestion
        reverseEscrow = channelEscrowAddresses[receivedPacket.destChannel]
        bank.MintCoins(reverseEscrow, token.denom, token.amount)
```

      }

>       // if it wasn't a source token on receive, then we simply had minted vouchers and burned them in the receive.

Reviewer 2: ```suggestion
      // if it wasn't a source token on receive: 
      //Previous Actions: 
            //1. Mint Vouchers 
            //2. Burn Vouchers in the receive
```

>       // So no state changes were made, and thus no reversion is necessary

Reviewer 2: ```suggestion
      //Reverse Actions: 
      // None. No state changes occurred and thus no reversion is necessary
```

```typescript
function onTimeoutPacket(packet: Packet) {
  // check if the packet sent is from a previously forwarded packet

>   prevPacket = privateStore.get(packetForwardPath(packet.sourcePort, packet.sourceChannel, packet.sequence))

Reviewer 6: I have an observation on the naming here from reading this quite fresh: 
It took me quite a bit of time and mental gymnastic to get the packets straight in my head here (to be fair, this could just be because I am not so used to this).
In particular, I found it tricky to map all these names to the right context:
- Forwarded packet
- Previous packet
- Sent packet
- Received packet
- Packet (in the context of onTimeout, onAck, OnRecv)

At first I thought it was an error that the timeout packet was sent in to `revertInFlightChanges` as sentPacket, because "surely the forwarded packet, aka prevPacket was the one we sent". But after rereading, I realized that forwardedPacket was not actually received packet. So it is correct, of course, but with all these names it was not so easy to make a mental model of the different packets.

Not sure what the best way to deal with this is, but perhaps there is a way we could tweak the names to make it a bit easier to grasp on first read.

Reviewer 5: This is good feedback; thanks, @gjermundgaraba. Maybe a diagram would also help; I can try to draw something in the next couple of days.

Author: Thank you for the feedback @gjermundgaraba . Will mull over better naming, but if you have some suggestions feel free to share!

}
```

```typescript
// revertInFlightChanges reverts the receive packet and send packet
// that occurs in the middle chains during a packet forwarding
// If an error occurs further down the line, the state changes
// on this chain must be reverted before sending back the error acknowledgement
// to ensure atomic packet forwarding
function revertInFlightChanges(sentPacket: Packet, receivedPacket: Packet) {

>   forwardEscrow = channelEscrowAddresses[sentPacket.sourceChannel]

Reviewer 5: Shouldn't this be `forwardEscrow = channelForwardingAddresses[receivedPacket.destChannel]` instead?
